applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_structural_response_function.cpp

```diff
@@ -0,0 +1,942 @@
+
+                //check for output variable to save later the computed sensitivity
+                if (KratosComponents<Variable<array_1d<double,3>>>::Has(output_label) == true)
+                {
+                    const Variable<array_1d<double,3>>& r_output_variable =
+                        KratosComponents<Variable<array_1d<double,3>>>::Get(output_label);
+                    this->UpdateConditionSensitivities(r_variable, r_output_variable);
+                }
+                else
+                    KRATOS_ERROR << "Unsupported condition variable for output: " << output_label << "." << std::endl;
+
+            }
+            else
+                KRATOS_ERROR << "Unsupported condition variable: " << label << "." << std::endl;
+        }
+
+        KRATOS_CATCH("");
+    }
+
+
+    /// Calculate the scalar valued response function
+    double AdjointStructuralResponseFunction::CalculateValue(ModelPart& rModelPart)
+    {
+        return 0.0;
+    }
+
+    template <typename TDataType>
+    void AdjointStructuralResponseFunction::UpdateNodalSensitivities(Variable<TDataType> const& rSensitivityVariable)
+    {
+        KRATOS_TRY;
+
+        ModelPart& r_model_part = this->GetModelPart();
+        ProcessInfo& r_process_info = r_model_part.GetProcessInfo();
+        //double delta_time = -r_process_info[DELTA_TIME];
+        const int num_threads = 1; //OpenMPUtils::GetNumThreads();
+        std::vector<Vector> sensitivity_vector(num_threads);
+        std::vector<Vector> response_gradient(num_threads);
+        std::vector<Vector> adjoint_vector(num_threads);
+        std::vector<Matrix> sensitivity_matrix(num_threads);
+
+        Communicator& r_comm = r_model_part.GetCommunicator();
+        if (r_comm.TotalProcesses() > 1)
+        {
+            // here we make sure we only add the old sensitivity once
+            // when we assemble.
+#pragma omp parallel
+            {
+                ModelPart::NodeIterator nodes_begin;
+                ModelPart::NodeIterator nodes_end;
+                OpenMPUtils::PartitionedIterators(r_model_part.Nodes(), nodes_begin, nodes_end);
+                for (auto it = nodes_begin; it != nodes_end; ++it)
+                    if (it->FastGetSolutionStepValue(PARTITION_INDEX) != r_comm.MyPID())
+                        it->FastGetSolutionStepValue(rSensitivityVariable) =
+                            rSensitivityVariable.Zero();
+            }
+        }
+        // Assemble element contributions.
+/*#pragma omp parallel
+        {
+            ModelPart::ElementIterator elements_begin;
+            ModelPart::ElementIterator elements_end;
+            OpenMPUtils::PartitionedIterators(r_model_part.Elements(),
+                                              elements_begin, elements_end);
+            int k = OpenMPUtils::ThisThread();*/
+
+            int k = 0;
+
+            //for (auto it = elements_begin; it != elements_end; ++it)
```

Comment:
From this point on I am completely lost in this function. Please clean it up and make smaller functions.